File: bank-kyc-system/src/main/java/com/bank/project/service/ManagerKycService.java
package com.bank.project.service;

import com.bank.project.dto.ManagerKycDecisionDTO;
import com.bank.project.entity.BankAccount;
import com.bank.project.entity.KycApplication;
import com.bank.project.entity.User;
import com.bank.project.enums.AccountStatus;
import com.bank.project.enums.AccountType;
import com.bank.project.enums.KycStatus;
import com.bank.project.repository.BankAccountRepository;
import com.bank.project.repository.KycApplicationRepository;
import com.bank.project.repository.UserRepository;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ManagerKycService {

    private final KycApplicationRepository kycRepo;
    private final UserRepository userRepository;
    private final BankAccountRepository bankAccountRepo;
    private final JavaMailSender mailSender;

    // ✅ Fetch only clerk-verified KYC
    public List<KycApplication> getPendingApplications() {
        return kycRepo.findByStatusAndClerkIsNotNullAndManagerIsNull(KycStatus.CLERK_VERIFIED);
    }

    public KycApplication getKycById(Long id) {
        return kycRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("KYC application not found"));
    }

    @Transactional
    public void reviewKyc(Long kycId, ManagerKycDecisionDTO dto) {
        KycApplication kyc = getKycById(kycId); // fetch the KYC record

        // ✅ Require remarks if rejected
        if (!dto.isApproved() && (dto.getRemarks() == null || dto.getRemarks().isEmpty())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Rejection reason is required");
        }

        // ✅ Fetch manager user entity
        User manager = userRepository.findById(dto.getManagerId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Manager not found"));

        kyc.setManager(manager);
        kyc.setRemarks(dto.getRemarks());
        kyc.setStatus(dto.isApproved() ? KycStatus.APPROVED : KycStatus.REJECTED);
      
        kycRepo.save(kyc);

        if (dto.isApproved()) {
        	createBankAccountAndNotify(kyc.getUser());
        }
    }

    // ✅ Generate bank account + send email
//    private void createBankAccountAndNotify(User user) {
//        Optional<BankAccount> existing = bankAccountRepo.findByUser(user);
//        if (existing.isPresent()) return; // avoid duplicate accounts
//
//        String accountNumber = "AC" + System.currentTimeMillis();
//        String branch = "Mumbai Branch";
//        String ifsc = "BANK0001234";
//        BigDecimal initialBalance = new BigDecimal("0.0000000");
//
//        BankAccount account = BankAccount.builder()
//                .accountNumber(accountNumber)
//                .accountType(user.getAccountType())
//                .status(AccountStatus.ACTIVE)
//                .branchName(branch)
//                .ifscCode(ifsc)   
//                .balance(initialBalance)
//                .user(user)
//                .build();
//
//        bankAccountRepo.save(account);
//
//        // 📧 Send account info email
//        SimpleMailMessage message = new SimpleMailMessage();
//        message.setTo(user.getEmail());
//        message.setSubject("Bank Account Approved");
//
//        message.setText("Dear " + user.getFirstName() + ",\n\n" +
//                "Congratulations! Your KYC is approved and your bank account has been created.\n\n" +
//                "Account Number: " + accountNumber + "\n" +
//                "Account Type: " + user.getAccountType() + "\n" +
//                "Account Status: ACTIVE\n" +
//                "IFSC Code: " + ifsc + "\n" +
//                "Branch: " + branch + "\n\n" +
//                "Thank you,\nBank Team");
//
//        mailSender.send(message);
//    }

    private void createBankAccountAndNotify(User user) {
        Optional<BankAccount> existing = bankAccountRepo.findByUser(user);
        if (existing.isPresent()) return; // Avoid duplicate accounts

        // ✅ Fetch latest APPROVED KYC for this user
        KycApplication approvedKyc = kycRepo.findTopByUserAndStatusOrderBySubmittedAtDesc(user, KycStatus.APPROVED)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "No approved KYC found for user"));

        String accountNumber = "AC" + System.currentTimeMillis();
        String branch = "Mumbai Branch";
        String ifsc = "BANK0001234";
        BigDecimal initialBalance = BigDecimal.ZERO;

        BankAccount account = BankAccount.builder()
                .accountNumber(accountNumber)
                .accountType(approvedKyc.getAccountType())   // ✅ Corrected
                .kycStatus(KycStatus.APPROVED)
                .status(AccountStatus.ACTIVE)
                .branchName(branch)
                .ifscCode(ifsc)
                .balance(initialBalance)
                .user(user)
                .build();

        bankAccountRepo.save(account);

        // 📧 Email notification
        SimpleMailMessage message = new SimpleMailMessage();
        message.setTo(user.getEmail());
        message.setSubject("Bank Account Approved");
        message.setText("Dear " + user.getFirstName() + ",\n\n" +
                "Congratulations! Your KYC is approved and your bank account has been created.\n\n" +
                "Account Number: " + accountNumber + "\n" +
                "Account Type: " + approvedKyc.getAccountType() + "\n" +
                "Account Status: ACTIVE\n" +
                "IFSC Code: " + ifsc + "\n" +
                "Branch: " + branch + "\n\n" +
                "Thank you,\nBank Team");

        mailSender.send(message);
    }

    // ✅ New: Get Approved KYC Applications
    public List<KycApplication> getApprovedKycs() {
        return kycRepo.findByStatus(KycStatus.APPROVED);
    }

    // ✅ New: Get Rejected KYC Applications
    public List<KycApplication> getRejectedKycs() {
        return kycRepo.findByStatus(KycStatus.REJECTED);
    }
}

File: BankKycSystemdotnet/BankKycSystem/Services/Implementations/ManagerKycService.cs
using BankKycSystem.Data;
using BankKycSystem.DTOs;
using BankKycSystem.Entities;
using BankKycSystem.Enums;
using BankKycSystem.Services.Interfaces;
using Microsoft.EntityFrameworkCore;
using System.Net;
using System.Net.Mail;

namespace BankKycSystem.Services.Implementations
{
    public class ManagerKycService : IManagerKycService
    {
        private readonly ApplicationDbContext _context;
        private readonly IConfiguration _configuration;

        public ManagerKycService(ApplicationDbContext context, IConfiguration configuration)
        {
            _context = context;
            _configuration = configuration;
        }

        public async Task<List<KycApplication>> GetPendingApplicationsAsync()
        {
            return await _context.KycApplications
                .Include(k => k.User)
                .Include(k => k.Clerk)
                .Where(k => k.Status == KycStatus.FORWARDED || k.Status == KycStatus.CLERK_VERIFIED)
                .ToListAsync();
        }

        public async Task<KycApplication?> GetKycByIdAsync(long id)
        {
            return await _context.KycApplications
                .Include(k => k.User)
                .Include(k => k.Clerk)
                .FirstOrDefaultAsync(k => k.Id == id);
        }

        public async Task ReviewKycAsync(long id, ManagerKycDecisionDTO dto, string managerEmail)
        {
            var application = await _context.KycApplications
                .Include(k => k.User)
                .FirstOrDefaultAsync(k => k.Id == id);

            if (application == null)
                throw new Exception("KYC application not found.");

            var manager = await _context.Users.FirstOrDefaultAsync(u => u.Email == managerEmail);
            if (manager == null)
                throw new Exception("Manager not found.");

            application.Manager = manager;
            application.Remarks = dto.Remarks;
            application.Status = dto.Approved ? KycStatus.APPROVED : KycStatus.REJECTED;

            if (dto.Approved)
            {
                await CreateBankAccountAndNotifyAsync(application.User!, application.AccountType);
            }

            await _context.SaveChangesAsync();
        }

        private async Task CreateBankAccountAndNotifyAsync(User user, AccountType? accountType)
        {
            // Avoid duplicate accounts
            if (await _context.BankAccounts.AnyAsync(b => b.UserId == user.Id))
                return;

            var accountNumber = GenerateAccountNumber();
            var branchName = "Main Branch";
            var ifsc = GenerateIfscCode();

            var account = new BankAccount
            {
                User = user,
                AccountNumber = accountNumber,
                AccountType = accountType,
                BranchName = branchName,
                IfscCode = ifsc,
                Balance = 0m,
                Status = AccountStatus.ACTIVE,
                KycStatus = KycStatus.APPROVED
            };

            _context.BankAccounts.Add(account);
            await _context.SaveChangesAsync();

            
            SendApprovalEmail(user.Email, user.FirstName, accountNumber, accountType, branchName, ifsc);
        }

        private void SendApprovalEmail(string toEmail, string firstName, string accountNumber, AccountType? accountType, string branchName, string ifsc)
        {
            var smtpHost = _configuration["Smtp:Host"];
            var smtpPort = int.Parse(_configuration["Smtp:Port"]);
            var smtpUser = _configuration["Smtp:Username"];
            var smtpPass = _configuration["Smtp:Password"];
            var fromEmail = _configuration["Smtp:FromEmail"];

            using (var client = new SmtpClient(smtpHost, smtpPort))
            {
                client.EnableSsl = true;
                client.Credentials = new NetworkCredential(smtpUser, smtpPass);

                var mail = new MailMessage
                {
                    From = new MailAddress(fromEmail, "Bank KYC System"),
                    Subject = "Bank Account Approved",
                    Body = $"Dear {firstName},\n\n" +
                           $"Congratulations! Your KYC is approved and your bank account has been created.\n\n" +
                           $"Account Number: {accountNumber}\n" +
                           $"Account Type: {accountType}\n" +
                           $"Account Status: ACTIVE\n" +
                           $"IFSC Code: {ifsc}\n" +
                           $"Branch: {branchName}\n\n" +
                           $"Thank you,\nBank Team",
                    IsBodyHtml = false
                };

                mail.To.Add(toEmail);
                client.Send(mail);
            }
        }

        public async Task<List<KycApplication>> GetApprovedKycsAsync()
        {
            return await _context.KycApplications
                .Include(k => k.User)
                .Where(k => k.Status == KycStatus.APPROVED)
                .ToListAsync();
        }

        public async Task<List<KycApplication>> GetRejectedKycsAsync()
        {
            return await _context.KycApplications
                .Include(k => k.User)
                .Where(k => k.Status == KycStatus.REJECTED)
                .ToListAsync();
        }

        private string GenerateAccountNumber()
        {
            var random = new Random();
            return string.Concat(Enumerable.Range(0, 12).Select(_ => random.Next(0, 10).ToString()));
        }

        private string GenerateIfscCode()
        {
            return "BANK" + new Random().Next(1000, 9999);
        }
    }
}
